libkatana_python_native/include/katana/python/TypeTraits.h

```diff
@@ -0,0 +1,79 @@
+#ifndef KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_TYPETRAITS_H_
+#define KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_TYPETRAITS_H_
+
+#include <pybind11/pybind11.h>
+
+namespace katana {
+
+template <typename T, typename Enable = void>
+struct PythonTypeTraits {};
+
+template <typename T>
+struct PythonTypeTraits<T, std::void_t<decltype(pybind11::type::of<T>())>> {
+  /// All pybind11 types use void* at the ctypes level.
+  static pybind11::object ctypes_type() {
+    pybind11::module_ ctypes = pybind11::module_::import("ctypes");
+    return ctypes.attr("c_void_p");
+  }
+
+  /// Get a Python object which represents the type in Python for purposes of
+  /// numba or type checking.
+  static pybind11::object representation() { return pybind11::type::of<T>(); }
+};
+
+#define PYTHON_TYPE_TRAITS(T, numpy_name, ctypes_name)                         \
+  template <>                                                                  \
+  struct PythonTypeTraits<T> {                                                 \
+    static constexpr const char* name = numpy_name;                            \
+    static pybind11::object default_dtype() {                                  \
+      pybind11::module_ numpy = pybind11::module_::import("numpy");            \
+      return numpy.attr(numpy_name);                                           \
+    }                                                                          \
+    static pybind11::object ctypes_type() {                                    \
+      pybind11::module_ ctypes = pybind11::module_::import("ctypes");          \
+      return ctypes.attr(ctypes_name);                                         \
+    }                                                                          \
+    static pybind11::object representation() { return ctypes_type(); }         \
+  }
+
+#define PYTHON_TYPE_TRAITS_BY_PREFIX(prefix)                                   \
+  PYTHON_TYPE_TRAITS(prefix##_t, #prefix, "c_" #prefix)
+
+PYTHON_TYPE_TRAITS_BY_PREFIX(uint8);
+PYTHON_TYPE_TRAITS_BY_PREFIX(uint16);
+PYTHON_TYPE_TRAITS_BY_PREFIX(uint32);
+PYTHON_TYPE_TRAITS_BY_PREFIX(uint64);
+
+PYTHON_TYPE_TRAITS_BY_PREFIX(int8);
+PYTHON_TYPE_TRAITS_BY_PREFIX(int16);
+PYTHON_TYPE_TRAITS_BY_PREFIX(int32);
+PYTHON_TYPE_TRAITS_BY_PREFIX(int64);
+
+PYTHON_TYPE_TRAITS(float, "float32", "c_float");
+PYTHON_TYPE_TRAITS(double, "float64", "c_double");
+
+#undef PYTHON_TYPE_TRAITS_BY_PREFIX
+#undef PYTHON_TYPE_TRAITS
+
+/// There is no numpy dtype for bool. Requesting it will give a C++ compiler
+/// error.
```

Comment:
Seriously? I thought numpy had a bool that was 1 byte?

Reply:
It does. But C++ `bool` is **not** always one byte. :-)